File: scala/P13.scala
// P13 Run-length encoding of a list (direct solution).

// Values factory
def makeValue[T](n: Int, x: T): Any =
  if (n == 1) x else (n, x)

// A simple solution, but there's still some kind of sublists
def f1: List[Any] => List[Any] = {
  case Nil => Nil
  case x :: xs =>
    val (a, b) = xs.span(x ==)
    makeValue(a.length + 1, x) :: f1(b)
}

// Recursion, "look-behind"
def f2: List[Any] => List[Any] = {
  case Nil => Nil
  case x :: xs => toValues(1, x, xs)
}

def toValues(n: Int, x: Any, l: List[Any]): List[Any] = l match {
  case Nil => List(makeValue(n, x))
  case y :: ys => if (y == x) toValues(n + 1, y, ys) else makeValue(n, x) :: toValues(1, y, ys)
}

// With custom data
abstract sealed class Entry[T]
case class Value[T](v: T) extends Entry[T]
case class Sequence[T](n: Int, v: T) extends Entry[T]

// Entries factory
def makeEntry[T](n: Int, x: T): Entry[T] =
  if (n == 1) Value(x) else Sequence(n, x)

// Using "span" 
def f3[T]: List[T] => List[Entry[T]] = {
  case Nil => Nil
  case x :: xs =>
    val (a, b) = xs.span(x ==)
    makeEntry(a.length + 1, x) :: f3(b)
}

// Direct method using custom data
def f4[T]: List[T] => List[Entry[T]] = {
  case Nil => Nil
  case x :: xs => toEntries(1, x, xs)
}

def toEntries[T](n: Int, x: T, l: List[T]): List[Entry[T]] = l match {
  case Nil => List(makeEntry(n, x))
  case y :: ys => if (y == x) toEntries(n + 1, y, ys) else makeEntry(n, x) :: toEntries(1, y, ys)
}

File: haskell/p13.hs
-- P13 Run-length encoding of a list (direct solution).

module P13 where

data Entry a = Value a | Sequence Int a deriving Show

-- Factory function
makeEntry :: Int -> a -> Entry a
makeEntry 1 a = Value a
makeEntry n a = Sequence n a

-- A simple solution, but there's still some kind of sublists
f1 :: Eq a => [a] -> [Entry a]
f1 [] = []
f1 (x:xs) = let (a, b) = span (== x) xs in
            makeEntry (succ $ length a) x : f1 b

-- Recursion, "look-ahead"
f2 :: Eq a => [a] -> [Entry a]
f2 [] = []
f2 xs = f 1 xs
    where f n [x] = [makeEntry n x]
          f n (a:bs@(b:_)) = if a == b then f (succ n) bs else makeEntry n a : f 1 bs

-- Recursion, "look-behind"
f3 :: Eq a => [a] -> [Entry a]
f3 [] = []
f3 (x:xs) = f 1 x xs
    where f n x [] = [makeEntry n x]
          f n x (y:ys) = if y == x then f (succ n) y ys else makeEntry n x : f 1 y ys